### ISSUE TITLE
torch.rand() causes non deterministic behavior

### ISSUE THREAD

**Issue by TParcollet** [speechbrain/speechbrain#2629]
### Describe the bug

Putting a line like `mask2 = torch.rand(5, requires_grad=False, device=self.device)` in the on_stage_start function on a recipe litteraly breaks determinism. Let's say that mask2 isn't used, at all, well just putting this line breaks determinism. I verified this by simply printing the sum of a parameter in the ASR model, and just calling this torch.rand() will make different value of the sum appear. This is very concerning. 

Very weird, if we create a generator with the hparams seed before and pass it to torch.rand(), then the problem disappears. Shouldn't torch.manual_seed() handle this? 

### Expected behaviour

calling torch.rand shouldn't alter the determinism of the model.

**Comment by asumagic** [comment]
Oof, I haven't tested it, but it seems like the seed for CPU and CUDA devices are separate, see [`torch.cuda.manual_seed`](https://pytorch.org/docs/stable/generated/torch.cuda.manual_seed.html) and [`torch.cuda.manual_seed_all`](https://pytorch.org/docs/stable/generated/torch.cuda.manual_seed_all.html#torch.cuda.manual_seed_all).

Might need a bit of thought behind how to configure this, e.g. when it comes to multi-GPU, as it might be undesirable for all GPUs to be using the same seed (e.g. if they end up with the same sequence of masks).

**Comment by asumagic** [comment]
Actually this might not be true, [`torch.manual_seed` says it affects all devices](https://pytorch.org/docs/stable/generated/torch.manual_seed.html), which would also be problematic for the reason state above. Needs more investigation.

**Comment by TParcollet** [comment]
There isn't much we can do about this except reforging SpeechBrain by using Generators. Luckily, the whole data pipeline is exempt of this, so the problem will only arise when rand operations are called on the model, which is, imho, very fine as it is unlikely that someone will call torch.rand() without integrating it to the graph (i.e. inducing a change anyway). We should talk about that during the next meeting for 1.1 still.
